src/librustdoc/clean/types.rs

```diff
@@ -1394,33 +1408,50 @@ impl Type {
     }
 }
 
-impl GetDefId for Type {
-    fn def_id(&self) -> Option<DefId> {
+impl Type {
+    fn inner_def_id(&self, cache: Option<&Cache>) -> Option<DefId> {
+        fn inner<T: GetDefId>(t: &T, cache: Option<&Cache>) -> Option<DefId> {
+            match cache {
+                Some(c) => t.def_id_full(c),
+                None => t.def_id(),
+            }
+        }
+
         match *self {
             ResolvedPath { did, .. } => Some(did),
-            Primitive(p) => cache().primitive_locations.get(&p).cloned(),
+            Primitive(p) => cache.and_then(|c| c.primitive_locations.get(&p).cloned()),
             BorrowedRef { type_: box Generic(..), .. } => {
-                Primitive(PrimitiveType::Reference).def_id()
+                inner(&Primitive(PrimitiveType::Reference), cache)
```

Comment:
Maybe not here, but it would be nice to avoid recursing here. You could do that by doing `let p: PrimitiveType = match { ... }` and returning early in the match if it isn't a primitive.